lib/sycamore/sycamore/lineage.py

```diff
@@ -106,3 +111,135 @@ def doc_to_name(doc: Document) -> str:
 
         assert isinstance(doc, Document)
         return doc.doc_id or str(uuid.uuid4())
+
+
+class AutoMaterialize(NodeTraverse):
+    """Automatically add materialize nodes after every node in an execution.
+
+    Usage:
+       from sycamore.lineage import AutoMaterialize
+       ctx = sycamore.init()
+       ctx.rewrite_rules.append(AutoMaterialize())
+
+       # override base directory for materialization
+       ctx.rewrite_rules.append(AutoMaterialize("/home/example/subdir"))
+
+       # override multiple parameters, root is the root directory for the materialize operations
+       # the remaining parameters work the same as docset.materialize(), and are simply passed through.
+       a = AutoMaterialize({"root":Path|str, fs=pyarrow.fs, name=fn, clean=bool, tobin=fn})
+       ctx.rewrite_rules.append(a)
+
+    Nodes in the plan will automatically be named. You can specify a name by defining it for the node:
+       ctx = sycamore.init()
+       ds = ctx.read.document(docs, materialize={"name": "reader"}).map(noop_fn, materialize={"name": "noop"})
+       # NOTE: names in a single execution must be unique. This is guaranteed by auto naming
+       # NOTE: automatic names are not guaranteed to be stable
+    """
+
+    def __init__(self, path: Union[str, Path, dict] = {}):
+        super().__init__()
+        if isinstance(path, str) or isinstance(path, Path):
+            path = {"root": path}
+        if "clean" not in path:
+            path["clean"] = True
+
+        self._path = path
+        self.directory = path.pop("root", None)
+        self._basename_to_count: dict[str, int] = {}
+
+    def once(self, context, node):
+        self._name_unique = set()
+        node = node.traverse(after=self._naming_pass())
+
+        self._setup_directory()
+
+        node = node.traverse(visit=self._cleanup_pass())
```

Comment:
what's the reason to separate naming and cleanup?

Reply:
Easier to understand code. I could do pass1 and pass2, but then you have to read the code to figure out what it is doing. naming_and_cleanup_pass would explain, but would mean the function isn't single purpose.

Since the performance loss is irrelevant, I prioritized clearer code.